sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PrivateFieldUsedAsLocalVariable.cs

```diff
@@ -36,340 +36,216 @@ namespace SonarAnalyzer.Rules.CSharp
-                {
-                    referenceSyntax = referenceSyntax.Parent;
-                }
-
-                if (referenceSyntax.Parent != null &&
-                    referenceSyntax.Parent.IsKind(SyntaxKind.ConditionalAccessExpression))
-                {
-                    referenceSyntax = referenceSyntax.Parent;
-                }
-
-
-                var enclosingSymbol = semanticModel.GetEnclosingSymbol(potentialReference.SpanStart);
-                if (!builder.ContainsKey(enclosingSymbol))
-                {
-                    builder.Add(enclosingSymbol, new Dictionary<SyntaxNode, ISymbol>());
-                }
+                this.semanticModel = semanticModel;
+                this.privateFields = privateFields;
 
-                builder[enclosingSymbol].Add(referenceSyntax, referencedSymbol);
+                privateFieldNames = privateFields.Keys.Select(s => s.Name).ToHashSet();
+                reads = new Dictionary<ISymbol, Dictionary<ISymbol, HashSet<StatementSyntax>>>();
+                writes = new Dictionary<ISymbol, Dictionary<ISymbol, HashSet<StatementSyntax>>>();
             }
 
-            return builder;
-        }
-
-        private static void ExcludePrivateFieldsBasedOnReferences(
-            IImmutableDictionary<ISymbol, PrivateField> privateFields,
-            IDictionary<ISymbol, IDictionary<SyntaxNode, ISymbol>> referencesByEnclosingSymbol,
-            ISet<ISymbol> classMethods)
-        {
-            var referencedAtLeastOnceFromClassMethod = new HashSet<ISymbol>();
-
-            foreach (var references in referencesByEnclosingSymbol)
+            public bool IsRemovableField(ISymbol fieldSymbol)
             {
-                if (!classMethods.Contains(references.Key))
-                {
-                    foreach (var reference in references.Value)
-                    {
-                        privateFields[reference.Value].Excluded = true;
-                    }
+                var writesByMethod = writes.GetValueOrDefault(fieldSymbol);
+                var readsByMethod = reads.GetValueOrDefault(fieldSymbol);
 
-                    continue;
+                // No methods overwrite the field value
+                if (writesByMethod == null)
+                {
+                    return false;
                 }
 
-                foreach (var reference in references.Value)
-                {
-                    referencedAtLeastOnceFromClassMethod.Add(reference.Value);
+                // A field is removable when no method read it, or all methods that read it, first overwrite it
+                return readsByMethod == null
+                    || readsByMethod.Keys.All(MethodOverwritesBeforeReading);
 
-                    if (!IsReferenceToSingleFieldValue(reference))
-                    {
-                        privateFields[reference.Value].Excluded = true;
-                    }
+                bool MethodOverwritesBeforeReading(ISymbol methodSymbol)

```

Comment:
If you are sure this is a method symbol why don't you use `IMethodSymbol`?

Reply:
It could be property accessor too... Any suggestion for a common name?